File: 0086-partition-list/0086-partition-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
         # separate the list into 2 distinct lists and link them afterwards.
         # p1, p2 traverses the list and hd1 and hd2 are the heads of two lists
            hd1=p1=ListNode(0)
            hd2=p2=ListNode(0)
            while head:
                if head.val<x:
                    p1.next=head
                    p1=p1.next
                else:
                    p2.next=head
                    p2=p2.next
                head=head.next
            #join the lists
            p2.next=None
            p1.next=hd2.next
            return hd1.next

File: 0086-partition-list/0086-partition-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        if (head == null) return head;
        
        // separate the list into 2 distinct lists and link them afterwards
        ListNode head1 = new ListNode(0);
        ListNode head2 = new ListNode(0);
        ListNode node = head;
        ListNode node1 = head1;
        ListNode node2 = head2;
        
        while (node != null) {
            if (node.val < x) {
                node1.next = node;
                node1 = node1.next;
            } else {
                node2.next = node;
                node2 = node2.next;
            }
            node = node.next;
        }

        node1.next = head2.next; // join lists
        node2.next = null;        
        return head1.next;
    }
}